receiver/k8sobjectsreceiver/receiver.go

```diff
@@ -49,22 +56,75 @@ func newReceiver(params receiver.CreateSettings, config *Config, consumer consum
 		return nil, err
 	}
 
-	return &k8sobjectsreceiver{
-		client:   client,
-		setting:  params,
-		consumer: consumer,
-		objects:  config.Objects,
-		obsrecv:  obsrecv,
-		mu:       sync.Mutex{},
-	}, nil
+	objReceiver := &k8sobjectsreceiver{
+		client:         client,
+		setting:        params,
+		logger:         params.Logger,
+		consumer:       consumer,
+		objects:        config.Objects,
+		obsrecv:        obsrecv,
+		mu:             sync.Mutex{},
+		leaderElection: config.LeaderElection,
+	}
+
+	if config.LeaderElection.Enabled {
+		if objReceiver.leaderElection.LockName == "" {
+			return nil, errors.New("luckName must not be empty if LeaderElection enabled")
+		}
+
+		objReceiver.leaderElectionClient, err = config.getClient()
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return objReceiver, nil
+}
+
+func (kr *k8sobjectsreceiver) startFunc(ctx context.Context) {
+	for _, object := range kr.objects {
+		kr.start(ctx, object)
+	}
+}
+
+func (kr *k8sobjectsreceiver) startInLeaderElectionMode(ctx context.Context) {
+	leaderLost := make(chan struct{})
+
+	kr.mu.Lock()
+	kr.stopperChanList = append(kr.stopperChanList, leaderLost)
+	kr.mu.Unlock()
+
+	lr, err := k8sconfig.NewLeaderElector(kr.leaderElection, kr.leaderElectionClient, func(_ context.Context) {
+		kr.logger.Info("this instance of the component was selected as the current leader",
+			zap.String("lock name", kr.leaderElection.LockName))
+
+		kr.startFunc(ctx)
+	},
+		func() {
+			kr.logger.Error("this instance of the component was previously the leader but was removed as such")
+			leaderLost <- struct{}{}
```

Comment:
Unclear how does it work given that the channel is closed in the Shutdown method. Can you please add a test covering start/shutdown callbacks?

Reply:
It's complicated to test the start/stop function of the leader elector, so I'll go ahead and explore it